### DISCUSSION TITLE
How do you extend Tycho OSGi Packaging Plugin timestamp logic?

### DISCUSSION THREAD

**Discussion by djreilly356** [eclipse-tycho/tycho#2255] [Q&A]
The documentation states that :The timestamp generation logic is extensible. The primary use case is to generate build version qualifier based on the timestamp of the last project commit." But it does not give any examples or clues about how to do this. Do you create a java class/jar? If so, what is it supposed to do? What input does it get? Where do you define it? What are the requirements? etc.

Details please.

And an example would be nice.

**Comment by laeubi** [comment]
May I ask if you really want to write your own provider or do you want to change the current default for timestamps (e.g. formatting etc.?)

**Reply by djreilly356** [reply]
I want to write my own provider. Simply reformatting the time stamp will not fix our problem. We are experiencing constant build failures due to timestamp differences that are not real. We need to control what timestamps are set on the jars based on whether or not there have been any changes to the code since the last build. We can figure out what changed, and we have a method to gather the desired timestamp info, but we need to know how to pass that to tycho.

**Reply by laeubi** [reply]
You probably want to use the jgit timestamp provider then as an example how to write such extension:
https://github.com/eclipse-tycho/tycho/tree/master/tycho-buildtimestamp-jgit

but be aware that the SPi is currently a bit in flux so it might sligthly change over time, so if you upgrade your tycho version you might need to update your extension as well.

**Comment by mfvalenta** [comment]
I'm am involved in the same project as the original poster. The issue I am having is that, although writing such a provider is easy by using the tycho-buildtimestamp-jgit as an example, the issue is how to build, test and install the plugin into Tycho. Is there any documentation available that describes these steps? That is, is there any documentation available that describes how a 3rd party can build and deploy a custom tycho plugin?

I have looked at https://github.com/eclipse-tycho/tycho/blob/master/CONTRIBUTING.md and have loaded the tycho code base and created a timestamp provider that compiles but, as I mentioned above, I haven't found anything on-line that describes the next steps.

On a side note, on the CONTRIBUTING page, in the manual steps. there is the following text that mentions a target but the target does not seem to exist in the load code. Was the name of the target changed?

"For Tycho only: Configure the target platform: Open the file tycho-bundles-target/tycho-bundles-target.target and click on Set as Target Platform in the upper right corner of the target definition editor."

**Reply by laeubi** [reply]
Here you can see how to configure your build with a custom provider:

First you need to build a simple jar that contains your code and add it as a dependency:

https://github.com/eclipse-m2e/m2e-core/blob/f5e9cdc1a0007b70985353dd5dd5afbaa6ea3b96/m2e-parent/pom.xml#L196-L200

then you can activate it to be used like this:

https://github.com/eclipse-m2e/m2e-core/blob/f5e9cdc1a0007b70985353dd5dd5afbaa6ea3b96/m2e-parent/pom.xml#L188

**Comment by mfvalenta** [comment]
Thanks for you quick response. I suspect you are assuming that I have a higher level of Maven/Tycho knowledge then I have. We recently upgraded our build from a pure PDE build to a Tycho build so Maven/Tycho is new to us. The step I am uncertain about is the "build a simple jar" step. How do I build the jar and how do I deploy it so that the pom you mentioned in your comment will be able to find it? Is there any documentation that describes that process for a 3rd party Tycho plugin? Do I need to use the mvn command line to do that or is there commands available in Eclipse?

It seems that install Maven plugins are in the .m2 folder. If I have the Tycho code loaded in Eclipse, is there an Eclipse command or build I can run that will build and install Tycho into my local .m2 folder so I can then test my code? Or is this something I also need to use the command line for?

**Reply by laeubi** [reply]
Basically you create a simple maven project as described here: 

https://maven.apache.org/guides/getting-started/#how-do-i-make-my-first-maven-project

The you add a dependency to Tycho like in the jgit provider:

https://github.com/eclipse-tycho/tycho/blob/master/tycho-buildtimestamp-jgit/pom.xml

... of course you can also copy the project. So you don't need a full Tycho setup to do this.

Then you can use `mvn clean install` to install it into your local repository and use it in a local build. Next step would then to deploy it so you can use it in your team, for example a nexus server is a usual choice here.

The details depend a bit on your use cases, your it infrastructure and so on, so its a bit hard to give exhaustive answers, but if it is crucial to your business you can find dedicated support options for Tycho here:
https://github.com/eclipse-tycho/tycho#getting-support

what is maybe a bit more efficient / faster than evaluate all options your self, especially if your are new to Maven **and** Tycho, because usually something like a dedicated timestamp provider is more for advanced users that already have some deeper experience and wanted to finetune things and not something one would start with as part of a migration.

**Comment by mfvalenta** [comment]
Thanks again. I think I get the gist of it now.
